jobserver/migrations/0003_remove_global_projectdeveloper.py

```diff
@@ -0,0 +1,32 @@
+from django.db import migrations
+
+from jobserver.authorization.roles import DeploymentAdministrator, ProjectDeveloper
+
+
+def change_user_role(apps, from_role, to_role):
+    User = apps.get_model("jobserver", "User")
+    for user in User.objects.filter(roles__contains=from_role):
+        user.roles.remove(from_role)
+        user.roles.add(to_role)
+        user.save()
+
+
+def replace_user_projectdeveloper_with_deploymentadministrator(apps, schema_editor):
+    change_user_role(apps, ProjectDeveloper, DeploymentAdministrator)
+
+
+def replace_user_deploymentadministrator_with_projectdeveloper(apps, schema_editor):
+    change_user_role(apps, DeploymentAdministrator, ProjectDeveloper)
+
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ("jobserver", "0002_remove_user_github_access_tokens"),
+    ]
+
+    operations = [
+        migrations.RunPython(
+            replace_user_projectdeveloper_with_deploymentadministrator,
+            reverse_code=replace_user_deploymentadministrator_with_projectdeveloper,
```

Comment:
Iain asked for my opinion on reversible migrations. I think in general we want all migrations to be reversible because we want to be able to easily move forward and backwards through different versions of our database. However, this can also be achieved using `reverse_code=migrations.RunPython.noop`.

If I've understood the behaviour correctly, this is running in a transaction. So if the code throws an exception, no change is made and the deployment is stopped. We have no need to rollback at that point. Therefore, I don't think this is particularly useful from the point of view of production (I could be wrong though). However, we might want to rollback locally to move between different versions of the app and in this case, I think we do want to change the user roles back rather than doing a `noop`. For example, to spot and fix the error above (`add` v `append`) I needed to change the data back in order to execute the broken line of code, merely relying on `noop` to move between migrations wasn't enough.

That said, I don't think we should be keeping migrations around for very long, so that makes its use quite limited. In this case though, it's a very small overhead and I found it quite handy in order to be able to fix that bug.

Reply:
Thanks, in future I will consider the reversibility of any migration I write!

Reply:
That's really useful, thanks @lucyb.